### PR TITLE
[#3608] Verify that auto-provisioned device ids match device id pattern

### PR DESCRIPTION
**Pull request by harism**
This is initial implementation idea I had and thought it should cover all the paths device is created to prevent creating any deivce with id which does not conform to the device id pattern.

I think I need to look at integration tests also but need to do that bit later.

### CODE REVIEW

**Comment by harism** [comment]
I added integration test for `assertRegistration` to verify that assertion fails with invalid device id. But do you consider this is enough for integration tests and rely on it that all adapters do call `assertRegistration` and this will cover all adapter uses where using via gateway it has been possible to have invalid device id as it has been defined?

If not I will continue writing per adapter integration tests in a similar fashion too.

File: services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/app/ManagementServicesProducer.java [hunk 1]
```diff
@@ -55,6 +56,9 @@ public class ManagementServicesProducer {
     @Inject
     MongoDbBasedCredentialsConfigProperties credentialsServiceProperties;
 
+    @Inject
+    MongoDbBasedHttpServiceConfigOptions mongoDbBasedHttpServiceConfigOptions;
+
```
Comment (sophokles73) [inline services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/app/ManagementServicesProducer.java:61]: Now that these options are needed in multiple places, I'd rather have a corresponding producer in `org.eclipse.hono.deviceregistry.mongodb.app.ConfigPropertiesProducer` and adapt `HttpServerFactory` accordingly. You can then add a corresponding parameter to this class' `deviceManagementService()` method.

Comment (sophokles73) [reply services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/app/ManagementServicesProducer.java:61]: I believe you can do the same for the JDBC registry correspondingly.

**Review by sophokles73 — CHANGES_REQUESTED** [review]

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java [hunk 2]
```diff
@@ -276,6 +284,15 @@ public final Future<OperationResult<Id>> createDevice(
         Objects.requireNonNull(span);
 
         final String deviceIdValue = deviceId.orElseGet(() -> generateDeviceId(tenantId));
+        final Matcher matcher = Objects.requireNonNull(serviceConfig.getDeviceIdPattern()).matcher(deviceIdValue);
```
[the remark this answers is not in the document]
Comment (harism) [reply services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java]: Done.

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java [hunk 3]
```diff
@@ -276,6 +284,15 @@ public final Future<OperationResult<Id>> createDevice(
...
+
+        if (!matcher.matches()) {
+            return Future.failedFuture(new ClientErrorException(
+                    tenantId,
+                    HttpURLConnection.HTTP_BAD_REQUEST,
+                    "invalid device id \"" + deviceIdValue + "\""
```
[the remark this answers is not in the document]
Comment (harism) [reply services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java]: Done.

File: services/device-registry-base/src/test/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementServiceTest.java [hunk 4]
```diff
@@ -67,7 +70,10 @@ void setUp() {
         eventBus = mock(EventBus.class);
         final Vertx vertx = mock(Vertx.class);
         when(vertx.eventBus()).thenReturn(eventBus);
-        deviceManagementService = new TestDeviceManagementService(vertx);
+        final ServiceConfigProperties serviceConfig = mock(ServiceConfigProperties.class);
+        when(serviceConfig.getDeviceIdPattern())
+                .thenReturn(Pattern.compile(RegistryManagementConstants.DEFAULT_REGEX_DEVICE_ID));
```
[the remark this answers is not in the document]
Comment (harism) [reply services/device-registry-base/src/test/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementServiceTest.java]: Done.

File: services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/app/ManagementServicesProducer.java (hunk 1 above)
Comment (harism) [reply services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/app/ManagementServicesProducer.java:61]: Ok, I get back to this a bit later on better time.

Comment (harism) [reply services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/app/ManagementServicesProducer.java]: Done, hope I understood it correctly what you were requesting to change.

Comment (sophokles73) [reply services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/app/ManagementServicesProducer.java]: You did :-)

**Review by sophokles73 — APPROVED** [review]
LGTM

thanks for fixing :+1:

**Pull request merged**